go/consensus/tendermint/abci/mux.go

```diff
@@ -758,23 +760,27 @@ func (mux *abciMux) DeliverTx(req types.RequestDeliverTx) types.ResponseDeliverT
 }
 
 func (mux *abciMux) EndBlock(req types.RequestEndBlock) types.ResponseEndBlock {
+	// Use cached results when available.
+	if !mux.state.proposal.needsExecution() {
+		if len(mux.state.proposal.resultsDeliverTx) != 0 {
```

Comment:
This is a sanity check, why would you remove it?

Reply:
I thought you should remove whole `if` clause, as it looks like you copied this from the `DeliverTx` method.

Reply:
No, this is slightly different, it is making sure that CometBFT has called DeliverTx for all of the transactions and nothing has been left over (e.g. all results have been poped). In CometBFT 0.38+ such a hack will no longer be needed as BeginBlock/DeliverTx/EndBlock is replaced with a single FinalizeBlock so we can make sure we iterate over all transactions and even get proper transaction indices.

Reply:
> DeliverTx is called once for each transaction in the block.

Sanity check is OK then. I immediately thought this is was a copy-paste because the panic message was the same.